### PR TITLE
PYTHON-808: Include hash of result set metadata in prepared stmt id

### PR DESCRIPTION
**Pull request by mambocab**
PR for [PYTHON-808](https://datastax-oss.atlassian.net/browse/PYTHON-808) -- not ready to merge, waiting on [CASSANDRA-10786](https://issues.apache.org/jira/browse/CASSANDRA-10786).

### CODE REVIEW

**Comment by mambocab** [comment]
Rebasing `db90e21` onto `master`. This fixes some test bugs and brings in the new keyspace-per-query changes.

File: cassandra/cluster.py [hunk 1]
```diff
@@ -3771,8 +3772,11 @@ def _execute_after_prepare(self, host, connection, pool, response):
                 if self.prepared_statement:
                     # result metadata is the only thing that could have
                     # changed from an alter
-                    _, _, _, result_metadata = response.results
-                    self.prepared_statement.result_metadata = result_metadata
+                    (_, _, _,
+                     self.prepared_statement.result_metadata,
+                     new_metadata_id) = response.results
```
Comment (aboudreault) [inline cassandra/cluster.py]: That is becoming a bit messy. We should probably create a follow-up ticket to create a better structure for response results.

File: cassandra/protocol.py [hunk 2]
```diff
@@ -682,9 +686,14 @@ def recv_results_metadata(cls, f, user_type_map):
         else:
             paging_state = None
 
+        if flags & 0x0008:
```
Comment (aboudreault) [inline cassandra/protocol.py]: better to create a constant var rather than using the hexa number directly here

File: cassandra/cluster.py (hunk 1 above)
Comment (mambocab) [reply cassandra/cluster.py]: We've already got a ticket for it, actually: [PYTHON-601](https://datastax-oss.atlassian.net/browse/PYTHON-601)

File: cassandra/protocol.py (hunk 2 above)
Comment (mambocab) [reply cassandra/protocol.py]: Yep, thanks for catching this

**Comment by mambocab** [comment]
rebasing 57fe8fb onto master to deal with merge conflicts

File: cassandra/protocol.py (hunk 2 above)
Comment (mambocab) [reply cassandra/protocol.py]: Addressed :+1:

**Comment by ifesdjeen** [comment]
Can we get this one pulled in and committed to the `cassandra-test` branch, as otherwise Cassandra dtests will be failing.

**Comment by mambocab** [comment]
Squashing down 3759324 to a smaller commit history.

**Pull request merged**